src/Claim.ts

```diff
@@ -1,14 +1,72 @@
 /* tslint:disable:no-relative-imports */
-import * as bitcore from 'bitcore-lib'
 import * as crypto from 'crypto'
-import { canonize } from 'jsonld'
+import * as cuid from 'cuid'
+import * as jsonld from 'jsonld'
+import * as jsig from 'jsonld-signatures'
 
 import { IllegalArgumentException } from './Exceptions'
-import { Claim, ClaimAttributes, ClaimType, ClaimContext, isClaim } from './Interfaces'
+import {
+  Claim,
+  ClaimAttributes,
+  ClaimType,
+  ClaimContext,
+  DefaultClaimContext,
+  claimTypeDefaults,
+  isClaim,
+} from './Interfaces'
+import { DataParser } from './util/DataParser'
 
-export const canonizeClaim = async (claim: Claim): Promise<string> => {
-  const contextualClaim = { ...ClaimContext, ...claim }
-  return canonize(contextualClaim)
+DataParser(jsonld)
+jsig.use('jsonld', jsonld)
+
+const canonizeClaim = async (document: Claim): Promise<string> => {
+  const contextualClaim = {
+    '@context': {
+      ...DefaultClaimContext['@context'],
+      ...claimTypeDefaults[document.type]['@context'],
+      ...document['@context'],
+    },
+    type: document.type,
+    issuer: document.issuer,
+    issuanceDate: document.issuanceDate,
+    claim: document.claim,
+  }
+  return jsonld.canonize(contextualClaim)
+}
+
+/*
+  signClaim = (signingOptions: any) => async (document: Claim): Promise<Claim>
+    signingOptions:
+      algorithm: 'Ed25519VerificationKey2018'
+      creator: 'data:,<base58PublicKeyValue>'
+      privateKeyBase58: base58 rendition of an Ed22159 private key
+    document: any valid Claim
+  NOTE: If you choose to use a different signing algorhithm, the other signingOption keys may differ.
+  Refer to the jsonld-signatures library for more detail
+*/
+export const signClaim = (signingOptions: any) => async (document: Claim): Promise<Claim> => {
```

Comment:
Id like to see the errors extracted from the logic of signClaim, something like, what do you guys think?

```ts
const signClaimValidateClaimId = async (claim: Claim) => {
  if (!claim.id) throw new IllegalArgumentException('Cannot sign a claim that has an empty .id field')
  if (await getClaimId(claim) !== claim.id) throw new IllegalArgumentException('Cannot sign a claim whose id has been altered or generated incorrectly.')
}

const signClaimValidateSigningOptions = async (signingOptions: any) => {
  if (signingOptions.creater === null || signingOptions.creator === '')
    throw new IllegalArgumentException('Cannot sign a claim with an invalid creator in the signing options.')
}

const signClaimValidateArgs = async (signingOptions: any, claim: Claim) => {
  await signClaimValidateClaimId(claim)
  await signClaimValidateSigningOptions(signingOptions)
}

/*
  signClaim = (signingOptions: any) => async (document: Claim): Promise<Claim>
    signingOptions:
      algorithm: 'Ed25519VerificationKey2018'
      creator: 'data:,<base58PublicKeyValue>'
      privateKeyBase58: base58 rendition of an Ed22159 private key
    document: any valid Claim
  NOTE: If you choose to use a different signing algorhithm, the other signingOption keys may differ.
  Refer to the jsonld-signatures library for more detail
*/
export const signClaim = (signingOptions: any) => async (document: Claim): Promise<Claim> => {
  await signClaimValidateArgs(signingOptions, document)

  return await jsig.sign(
    {
      '@context': document['@context'],
      type: document.type,
      issuer: document.issuer,
      issuanceDate: document.issuanceDate,
      claim: document.claim,
    },
    signingOptions
  )
}
```

Reply:
That works for me :)

Reply:
Done